src/rdf4cpp/util/CharMatcher.cpp

```diff
@@ -0,0 +1,155 @@
+    template<class D, class Func, typename T = hwy::HWY_NAMESPACE::TFromD<D>>
+    void Foreach(D d, T const *HWY_RESTRICT in, size_t const count, hwy::HWY_NAMESPACE::Vec<D> const no,
+                 Func const &func) {
+        size_t const N = Lanes(d);
+
+        size_t idx = 0;
+        if (count >= N) {
+            for (; idx <= count - N; idx += N) {
+                hwy::HWY_NAMESPACE::Vec<D> const v = LoadU(d, in + idx);
+                if (!func(d, v)) {
+                    return;
+                }
+            }
+        }
+
+        // `count` was a multiple of the vector length `N`: already done.
+        if (HWY_UNLIKELY(idx == count))
+            return;
+
+        size_t const remaining = count - idx;
+        HWY_DASSERT(0 != remaining && remaining < N);
+        hwy::HWY_NAMESPACE::Vec<D> const v = LoadNOr(no, d, in + idx, remaining);
+        func(d, v);
+    }
+
+    std::optional<bool> try_match_simd_impl(std::string_view data, std::array<CharRange, 3> const &ranges, std::string_view single) {
+        // register usage:
+        // 1 - data
+        // 6 - ranges
+        // 5 - logic
+        // ? - singles
+        // 16 overall
+        bool found_unicode = false;
+        bool r = true;
+        using D = hwy::HWY_NAMESPACE::ScalableTag<int8_t>;  //NOLINT
+        using V = hwy::HWY_NAMESPACE::VFromD<D>;             //NOLINT
+        D d;
+        bool r0a = ranges[0].first < ranges[0].last;
+        V zero = r0a ? hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(ranges[0].first)) : hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(single.at(0)));
+        V r0b = hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(ranges[0].first - 1));
+        V r0e = hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(ranges[0].last + 1));
+        bool r1a = ranges[1].first < ranges[1].last;
+        V r1b = hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(ranges[1].first - 1));
+        V r1e = hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(ranges[1].last + 1));
+        bool r2a = ranges[2].first < ranges[2].last;
+        V r2b = hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(ranges[2].first - 1));
+        V r2e = hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(ranges[2].last + 1));
+        V unicode_bit = hwy::HWY_NAMESPACE::Set(d, 7);
+
+        // if this gets ported to variable size vectors, this needs to be changed (like above)
+        // this array prevents keeping in registers, but there are not enough registers anyway
+        std::array<V, simd_max_single_chars> si{};
+        size_t si_num = single.size();
+        if (si_num > si.size()) {
+            assert(false);
+            return std::nullopt;
+        }
+        for (size_t i = 0; i < si_num; ++i) {
+            si[i] = hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(single[i]));
+        }
+
+        Foreach(d, reinterpret_cast<int8_t const *>(data.data()), data.size(), zero, [&](auto d, auto in_vec) HWY_ATTR {
```

Comment:
Is foreach just a badly named `fold`? I'm confused what the `zero` param does

Reply:
zero is the placeholder that gets used, if there is no full vector of data left